pkg/apis/pipeline/v1beta1/pipeline_validation_test.go

```diff
@@ -3574,3 +3575,73 @@ func enableFeatures(t *testing.T, features []string) func(context.Context) conte
 		return s.ToContext(ctx)
 	}
 }
+
+func TestPipelineDeprecationWarning(t *testing.T) {
+	tests := []struct {
+		name          string
+		pipelineSpec  *PipelineSpec
+		expectedError *apis.FieldError
+	}{{
+		name: "Resources",
+		pipelineSpec: &PipelineSpec{
+			Tasks: []PipelineTask{{
+				Name: "non-final-task", TaskRef: &TaskRef{Name: "foo"},
+			}},
+			Resources: []PipelineDeclaredResource{{
+				Name:     "app-repo",
+				Type:     "git",
+				Optional: false,
+			}, {
+				Name:     "app-image",
+				Type:     "git",
+				Optional: false,
+			}},
+		},
+		expectedError: &apis.FieldError{
+			Message: "Resources field is deprecated in v1 Pipeline",
+			Paths:   []string{"Resources"},
+		},
+	}, {
+		name: "Bundle",
+		pipelineSpec: &PipelineSpec{
+			Tasks: []PipelineTask{{
+				Name: "non-final-task",
+				TaskRef: &TaskRef{
+					Name:   "foo",
+					Bundle: "test-bundle",
+				},
+			}},
+		},
+		expectedError: &apis.FieldError{
+			Message: "Bundle field is deprecated in v1 Pipeline",
+			Paths:   []string{"Bundle"},
+		},
+	}}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			ts := tt.pipelineSpec
+			ctx := context.Background()
+			err := ts.Validate(enableTektonOCIBundles(t)(ctx))
+			if err == nil && tt.expectedError.Error() != "" {
+				t.Fatalf("Expected an error, got nothing for %v", ts)
+			}
+			if d := cmp.Diff(tt.expectedError.Error(), err.Error(), cmpopts.IgnoreUnexported(apis.FieldError{})); d != "" {
+				t.Errorf("TaskSpec.Validate() errors diff %s", diff.PrintWantGot(d))
+			}
+		})
+	}
+}
+
+// TODO refactor in #5111
+func enableTektonOCIBundles(t *testing.T) func(context.Context) context.Context {
```

Comment:
might be easier to start with a single deprecated field such as timeout. Also, this function is defined elsewhere and should be imported instead of copy/pasted.

Reply:
This is because in pipeline_validation_test.go the package is v1beta1, while elsewhere in pipelineref_validation_test.go it is in v1beta1_test. I put the TODO comment as to resolve this in https://github.com/tektoncd/pipeline/issues/5111 if that makes sense to refactor in a separate PR.

Reply:
you can move it to https://github.com/tektoncd/pipeline/blob/main/pkg/apis/config/feature_flags.go-- that's where we put `EnableAlphaAPIFields` after a similar discussion